ingest/sqs/pools/transformer/pool_transformer.go

```diff
@@ -230,51 +209,138 @@ func (pi *poolTransformer) convertPool(
+	poolLiquidityCapUSDC, poolLiquidityCapUSDCErrorStr := pi.computeUSDCPoolLiquidityCapFromUOSMO(ctx, poolLiquidityCapUOSMO)
+
+	// Join error strings for pool liquidity cap.
+	poolLiquidityCapErrorStr = strings.Join([]string{poolLiquidityCapErrorStr, poolLiquidityCapUSDCErrorStr}, " ")
+
+	// Get pool denoms. Although these can be inferred from balances, this is safer.
+	// If we used balances, for pools with no liquidity, we would not be able to get the denoms.
+	denoms, err := pi.poolManagerKeeper.RouteGetPoolDenoms(ctx, pool.GetId())
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort denoms for consistent ordering.
+	sort.Strings(denoms)
+
+	// Mutates denomPairToTakerFeeMap with the taker fee for every uniquer denom pair in the denoms list.
+	err = retrieveTakerFeeToMapIfNotExists(ctx, denoms, denomPairToTakerFeeMap, pi.poolManagerKeeper)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the tick model for concentrated pools
+	var tickModel *sqsdomain.TickModel
+
+	// For CL pools, get the tick data
+	if pool.GetType() == poolmanagertypes.Concentrated {
+		tickData, currentTickIndex, err := pi.concentratedKeeper.GetTickLiquidityForFullRange(ctx, pool.GetId())
+		// If there is no error, we set the tick model
+		if err == nil {
+			tickModel = &sqsdomain.TickModel{
+				Ticks:            tickData,
+				CurrentTickIndex: currentTickIndex,
+			}
+			// If there is no liquidity, we set the tick model to nil and update no liquidity flag
+		} else if errors.Is(err, concentratedtypes.RanOutOfTicksForPoolError{PoolId: pool.GetId()}) {
+			tickModel = &sqsdomain.TickModel{
+				Ticks:            []queryproto.LiquidityDepthWithRange{},
+				CurrentTickIndex: -1,
+				HasNoLiquidity:   true,
+			}
+
+			// On any other error, we return the error
+		} else {
+			return nil, err
+		}
+	}
+
+	return &sqsdomain.PoolWrapper{
+		ChainModel: pool,
+		SQSModel: sqsdomain.SQSPool{
+			PoolLiquidityCap:      poolLiquidityCapUSDC,
+			PoolLiquidityCapError: poolLiquidityCapErrorStr,
+			Balances:              filteredBalances,
+			PoolDenoms:            denoms,
+			SpreadFactor:          spreadFactor,
+		},
+		TickModel: tickModel,
+	}, nil
+}
+
+// getPoolDenomsMap converts pool denoms to a map for faster lookup.
+func getPoolDenomsMap(poolDenoms []string) map[string]struct{} {
+	poolDenomsMap := make(map[string]struct{}, len(poolDenoms))
+
+	// Convert pool denoms to map
+	for _, poolDenom := range poolDenoms {
+		poolDenomsMap[poolDenom] = struct{}{}
+	}
+	return poolDenomsMap
+}
+
+// filterBalances filters out balances that do not belong to the pool.
+// Note that there are edge cases where gamm shares or some random
+// garbage tokens are in the balance that do not belong to the pool.
+// A mainnet example is pool ID 2 with the following extra denoms:
+// ibc/65BCD5909ED3D9E6223529017BC828ECBECCBE3F63D444EC44CE7412EF8C82D6
+// ibc/778F0504E33BBB66D0950FE12E29BA81C258ED0A10CCEF9CB0096BA9E22C5D61
```

Comment:
Is this some historical chain side logic that is no longer possible? I feel like this being possible now would break some important invariants..

Reply:
It is possible but the logic is equivalent to what used to be before. We still need to filter those out.

I simply extracted this into a separate function for readability and testability